articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+is not needed to dynamically query based on node labels, relationship types and parameters.
+
+It is also important to remember that not all injection attacks rely on quotation escaping of string literals,
+for example the following query, for a number literal, is still vulnerable when not using parameterization.
+
+`query = "MATCH (n) WHERE n.id =" + userid + ";"`
+
+Where the malicious userid input is: `"1 OR 1 = 1 WITH DISTINCT true AS haxxored MATCH (all) DETACH DELETE all; //"`
+
+=== Validation and Sanitization common Exploits
+
+Sanitization can also be used as a technique to clean up user input.
+Another way of keeping the input safe and clean is to use validation.
+Validation checks the input and makes sure it meets a set of certain criteria and will reject the input if it does not,
+in comparison to sanitization which cleans the input only.
+Validation can be used alongside sanitization, keep in mind that both techniques come with risks.
+
+=== Whitespace checks
+
+Checking user input for whitespace sounds like a good way to avoid injection, and in some cases it would work,
+consider the example:
+
+`"Robby' MATCH (s:Student) DETACH DELETE s //"`
+
+A validation check for whitespace would flag this query as invalid, but checking for whitespace alone isn’t enough.
+In Cypher using block comments to replace whitespaces is also valid, the following query would, therefore, pass whitespace validation checks:
+
+`"Robby'/&#42;&#42;/MATCH/&#42;&#42;/(s:Student)/&#42;&#42;/DETACH/&#42;&#42;/DELETE/&#42;&#42;/s/&#42;&#42;///"`
+
+Note that in this case, filtering for `/&#42;&#42;/` is still not enough as block comments can, themselves, contain random ignorable characters: `/&#42;&#42;thisisacomment&#42;&#42;/`.
+
+Checking for and cleaning up whitespace may be useful for your application, but shouldn't be relied upon as a secure way of avoiding Cypher injection.
+
+=== Unicode Encoding
+
+Another common exploit around input validation and sanitization is unicode encoding.
+Unicode encoding is where characters are encoded into their unicode equivalent.
+For example; the single quote character `'` can be encoded as `\u0027`.
+When sanitizing a string for the removal of escape quotation characters, it is important to also check for the unicode equivalent.
+The following query doesn't look like it escapes the string at first glance:
+
+`"Robby\u0027 MATCH (s:Student) DETACH DELETE s //"`
+
+But in reality, Cypher will resolve the unicode into a single quote and treat it as such in the compilation of the query.
+
+As mentioned earlier, malicious users may double encode their unicode input by additionally encoding the backslash, `\`, and `u`,
+which allows that to be resolved first in your client application code, and then resolved as the actual attack in the database.
+For example; `\u005C\u00750027` may be sent to the database as the input `\u0027`, thus successfully escaping the single quotation.
```

Comment:
Perhaps the section described earlier could be merged into this one, as explaining the same idea twice is a little repetitious?

Reply:
Fair enough :)